File: frontendios/src/services/wahooService.ts
/**
 * Service Wahoo OAuth pour EdgeCoach iOS
 * Gestion de l'intégration Wahoo (connexion, synchronisation)
 */

import { Linking } from 'react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import apiService from './api';

// Configuration Wahoo
const WAHOO_CONFIG = {
  CLIENT_ID: 'aSnKputUgUtaCxtVlAImFyp-EyYQQoFUmBZnCyTC1lM',
  REDIRECT_URI: 'edgecoach://auth/wahoo/callback',
  SCOPES: 'email user_read workouts_read offline_data',
  AUTH_URL: 'https://api.wahooligan.com/oauth/authorize',
  TOKEN_URL: 'https://api.wahooligan.com/oauth/token',
  API_URL: 'https://api.wahooligan.com/v1',
};

// Types
export interface WahooTokens {
  access_token: string;
  refresh_token: string;
  expires_in: number;
  token_type: string;
}

export interface WahooProfile {
  id: number;
  email: string;
  name: string;
  first_name?: string;
  last_name?: string;
  created_at?: string;
}

export interface WahooWorkout {
  id: number;
  name: string;
  sport: string;
  duration_seconds: number;
  distance_meters?: number;
  calories?: number;
  avg_heart_rate?: number;
  max_heart_rate?: number;
  created_at: string;
}

export interface WahooConnectionStatus {
  isConnected: boolean;
  profile?: WahooProfile;
  lastSync?: string;
  error?: string;
}

export interface WahooAuthResult {
  success: boolean;
  authUrl?: string;
  instructions?: string[];
  error?: string;
}

export interface WahooExchangeResult {
  success: boolean;
  tokens?: WahooTokens;
  profile?: WahooProfile;
  message?: string;
  error?: string;
}

class WahooService {
  private storageKeys = {
    STATE: 'wahoo_oauth_state',
    USER_ID: 'wahoo_oauth_user_id',
    TOKENS: 'wahoo_tokens',
    PROFILE: 'wahoo_profile',
  };

  /**
   * Générer l'URL d'autorisation Wahoo
   */
  generateAuthUrl(userId: string): string {
    const stateToken = `wahoo_${userId}_${Math.floor(Date.now() / 1000)}`;

    const params = new URLSearchParams({
      client_id: WAHOO_CONFIG.CLIENT_ID,
      redirect_uri: WAHOO_CONFIG.REDIRECT_URI,
      response_type: 'code',
      scope: WAHOO_CONFIG.SCOPES,
      state: stateToken,
    });

    // Stocker l'état pour vérification ultérieure
    AsyncStorage.setItem(this.storageKeys.STATE, stateToken);
    AsyncStorage.setItem(this.storageKeys.USER_ID, userId);

    return `${WAHOO_CONFIG.AUTH_URL}?${params.toString()}`;
  }

  /**
   * Démarrer le processus OAuth Wahoo
   */
  async startOAuth(userId: string): Promise<WahooAuthResult> {
    try {
      const authUrl = this.generateAuthUrl(userId);

      // Ouvrir l'URL dans le navigateur
      const canOpen = await Linking.canOpenURL(authUrl);
      if (!canOpen) {
        return {
          success: false,
          error: "Impossible d'ouvrir le navigateur pour l'authentification Wahoo",
        };
      }

      await Linking.openURL(authUrl);

      return {
        success: true,
        authUrl,
        instructions: [
          '1. Connectez-vous avec vos identifiants Wahoo',
          "2. Cliquez sur \"Autoriser\" pour EdgeCoach",
          "3. Vous serez redirigé vers l'application",
        ],
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.message,
      };
    }
  }

  /**
   * Vérifier si une URL est un callback Wahoo valide
   */
  isWahooCallbackUrl(url: string): boolean {
    if (!url) return false;
    return url.startsWith('edgecoach://auth/wahoo/callback');
  }

  /**
   * Extraire le code d'autorisation d'une URL de callback
   */
  extractAuthCodeFromUrl(url: string): { code: string | null; state: string | null; error: string | null } {
    try {
      // Parser les query params manuellement (React Native n'a pas URL/URLSearchParams)
      const queryStart = url.indexOf('?');
      if (queryStart === -1) {
        return { code: null, state: null, error: null };
      }

      const queryString = url.substring(queryStart + 1);
      const params: Record<string, string> = {};

      queryString.split('&').forEach(pair => {
        const [key, value] = pair.split('=');
        if (key && value) {
          params[decodeURIComponent(key)] = decodeURIComponent(value);
        }
      });

      return {
        code: params['code'] || null,
        state: params['state'] || null,
        error: params['error'] || null,
      };
    } catch {
      return { code: null, state: null, error: null };
    }
  }

  /**
   * Gérer le callback OAuth (appelé quand l'app reçoit le deep link)
   */
  async handleCallback(callbackUrl: string): Promise<WahooExchangeResult> {
    try {
      const { code, state, error } = this.extractAuthCodeFromUrl(callbackUrl);

      if (error) {
        return {
          success: false,
          error: `Erreur OAuth: ${error}`,
        };
      }

      if (!code) {
        return {
          success: false,
          error: "Code d'autorisation non trouvé dans l'URL",
        };
      }

      // Vérifier l'état
      const storedState = await AsyncStorage.getItem(this.storageKeys.STATE);
      if (state !== storedState) {
        return {
          success: false,
          error: "État OAuth invalide. Veuillez réessayer.",
        };
      }

      // Récupérer l'ID utilisateur
      const userId = await AsyncStorage.getItem(this.storageKeys.USER_ID);
      if (!userId) {
        return {
          success: false,
          error: 'ID utilisateur non trouvé',
        };
      }

      // Échanger le code contre les tokens via le backend
      return this.exchangeCodeForTokens(code, userId);
    } catch (error: any) {
      return {
        success: false,
        error: error.message,
      };
    }
  }

  /**
   * Échanger le code d'autorisation contre des tokens via le backend
   */
  async exchangeCodeForTokens(code: string, userId: string): Promise<WahooExchangeResult> {
    try {
      const response = await apiService.post<{
        access_token: string;
        refresh_token: string;
        expires_in: number;
        token_type: string;
        profile?: WahooProfile;
      }>('/auth/wahoo/exchange', {
        code,
        user_id: userId,
        redirect_uri: WAHOO_CONFIG.REDIRECT_URI,
      });

      const tokens: WahooTokens = {
        access_token: response.access_token,
        refresh_token: response.refresh_token,
        expires_in: response.expires_in,
        token_type: response.token_type,
      };

      // Stocker les tokens
      await AsyncStorage.setItem(this.storageKeys.TOKENS, JSON.stringify(tokens));

      // Si le profil est retourné, le stocker
      if (response.profile) {
        await AsyncStorage.setItem(this.storageKeys.PROFILE, JSON.stringify(response.profile));
      }

      // Nettoyer les données temporaires
      await AsyncStorage.removeItem(this.storageKeys.STATE);
      await AsyncStorage.removeItem(this.storageKeys.USER_ID);

      return {
        success: true,
        tokens,
        profile: response.profile,
        message: 'Connexion Wahoo établie avec succès!',
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.message,
      };
    }
  }

  /**
   * Vérifier le statut de connexion Wahoo
   */
  async getConnectionStatus(userId: string): Promise<WahooConnectionStatus> {
    try {
      const response = await apiService.get<{
        connected: boolean;
        profile?: WahooProfile;
        last_sync?: string;
      }>(`/auth/wahoo/status`, { user_id: userId });

      return {
        isConnected: response.connected,
        profile: response.profile,
        lastSync: response.last_sync,
      };
    } catch (error: any) {
      // Essayer de lire depuis le stockage local
      const tokensStr = await AsyncStorage.getItem(this.storageKeys.TOKENS);
      const profileStr = await AsyncStorage.getItem(this.storageKeys.PROFILE);

      if (tokensStr) {
        return {
          isConnected: true,
          profile: profileStr ? JSON.parse(profileStr) : undefined,
        };
      }

      return {
        isConnected: false,
        error: error.message,
      };
    }
  }

  /**
   * Déconnecter Wahoo
   */
  async disconnect(userId: string): Promise<{ success: boolean; error?: string }> {
    try {
      await apiService.post<{ message: string }>('/auth/wahoo/disconnect', { user_id: userId });

      // Nettoyer le stockage local
      await AsyncStorage.removeItem(this.storageKeys.TOKENS);
      await AsyncStorage.removeItem(this.storageKeys.PROFILE);
      await AsyncStorage.removeItem(this.storageKeys.STATE);
      await AsyncStorage.removeItem(this.storageKeys.USER_ID);

      return { success: true };
    } catch (error: any) {
      return {
        success: false,
        error: error.message,
      };
    }
  }

  /**
   * Synchroniser les entraînements Wahoo
   */
  async syncWorkouts(userId: string): Promise<{ success: boolean; workouts?: WahooWorkout[]; error?: string }> {
    try {
      const response = await apiService.post<{ workouts: WahooWorkout[]; count: number }>(
        '/auth/wahoo/sync',
        { user_id: userId }
      );

      return {
        success: true,
        workouts: response.workouts,
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.message,
      };
    }
  }

  /**
   * Récupérer les entraînements récents
   */
  async getRecentWorkouts(
    userId: string,
    limit: number = 10
  ): Promise<{ success: boolean; workouts?: WahooWorkout[]; error?: string }> {
    try {
      const response = await apiService.get<{ workouts: WahooWorkout[] }>('/auth/wahoo/workouts', {
        user_id: userId,
        limit: limit.toString(),
      });

      return {
        success: true,
        workouts: response.workouts,
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.message,
      };
    }
  }
}

export const wahooService = new WahooService();
export default wahooService;

File: EdgeCoach/Services/WahooService.swift
/**
 * Service Wahoo OAuth pour EdgeCoach iOS
 * Gestion de l'intégration Wahoo (connexion, synchronisation)
 * Aligné avec frontendios/src/services/wahooService.ts
 */

import Foundation
import UIKit

// MARK: - Wahoo Configuration

private enum WahooConfig {
    static let clientId = "aSnKputUgUtaCxtVlAImFyp-EyYQQoFUmBZnCyTC1lM"
    static let redirectUri = "edgecoach://auth/wahoo/callback"
    static let scopes = "email user_read workouts_read offline_data"
    static let authUrl = "https://api.wahooligan.com/oauth/authorize"
    static let tokenUrl = "https://api.wahooligan.com/oauth/token"
    static let apiUrl = "https://api.wahooligan.com/v1"
}

// MARK: - Wahoo Models

struct WahooTokens: Codable {
    let accessToken: String
    let refreshToken: String
    let expiresIn: Int
    let tokenType: String

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
        case expiresIn = "expires_in"
        case tokenType = "token_type"
    }
}

struct WahooProfile: Codable {
    let id: Int
    let email: String
    let name: String
    let firstName: String?
    let lastName: String?
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id, email, name
        case firstName = "first_name"
        case lastName = "last_name"
        case createdAt = "created_at"
    }
}

struct WahooWorkout: Codable, Identifiable {
    let id: Int
    let name: String
    let sport: String
    let durationSeconds: Double
    let distanceMeters: Double?
    let calories: Double?
    let avgHeartRate: Double?
    let maxHeartRate: Double?
    let createdAt: String

    enum CodingKeys: String, CodingKey {
        case id, name, sport
        case durationSeconds = "duration_seconds"
        case distanceMeters = "distance_meters"
        case calories
        case avgHeartRate = "avg_heart_rate"
        case maxHeartRate = "max_heart_rate"
        case createdAt = "created_at"
    }
}

struct WahooConnectionStatus {
    let isConnected: Bool
    let profile: WahooProfile?
    let lastSync: String?
    let error: String?
}

// MARK: - Storage Keys

private enum WahooStorageKeys {
    static let state = "wahoo_oauth_state"
    static let userId = "wahoo_oauth_user_id"
    static let tokens = "wahoo_tokens"
    static let profile = "wahoo_profile"
}

// MARK: - Wahoo Service

@MainActor
class WahooService {
    static let shared = WahooService()
    private let api = APIService.shared

    private init() {}

    // MARK: - Generate Auth URL

    func generateAuthUrl(userId: String) -> String {
        let stateToken = "wahoo_\(userId)_\(Int(Date().timeIntervalSince1970))"

        // Store state for verification
        UserDefaults.standard.set(stateToken, forKey: WahooStorageKeys.state)
        UserDefaults.standard.set(userId, forKey: WahooStorageKeys.userId)

        var components = URLComponents(string: WahooConfig.authUrl)!
        components.queryItems = [
            URLQueryItem(name: "client_id", value: WahooConfig.clientId),
            URLQueryItem(name: "redirect_uri", value: WahooConfig.redirectUri),
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "scope", value: WahooConfig.scopes),
            URLQueryItem(name: "state", value: stateToken)
        ]

        return components.url!.absoluteString
    }

    // MARK: - Start OAuth

    func startOAuth(userId: String) async -> (success: Bool, authUrl: String?, instructions: [String]?, error: String?) {
        let authUrl = generateAuthUrl(userId: userId)

        guard let url = URL(string: authUrl) else {
            return (false, nil, nil, "URL d'authentification invalide")
        }

        // Open URL in browser
        if await UIApplication.shared.canOpenURL(url) {
            await UIApplication.shared.open(url)

            return (
                true,
                authUrl,
                [
                    "1. Connectez-vous avec vos identifiants Wahoo",
                    "2. Cliquez sur \"Autoriser\" pour EdgeCoach",
                    "3. Vous serez redirigé vers l'application"
                ],
                nil
            )
        } else {
            return (false, nil, nil, "Impossible d'ouvrir le navigateur pour l'authentification Wahoo")
        }
    }

    // MARK: - Check Callback URL

    func isWahooCallbackUrl(_ url: URL) -> Bool {
        return url.absoluteString.hasPrefix("edgecoach://auth/wahoo/callback")
    }

    // MARK: - Extract Auth Code

    func extractAuthCodeFromUrl(_ url: URL) -> (code: String?, state: String?, error: String?) {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return (nil, nil, nil)
        }

        let code = components.queryItems?.first(where: { $0.name == "code" })?.value
        let state = components.queryItems?.first(where: { $0.name == "state" })?.value
        let error = components.queryItems?.first(where: { $0.name == "error" })?.value

        return (code, state, error)
    }

    // MARK: - Handle Callback

    func handleCallback(callbackUrl: URL) async -> (success: Bool, tokens: WahooTokens?, profile: WahooProfile?, message: String?, error: String?) {
        let (code, state, errorParam) = extractAuthCodeFromUrl(callbackUrl)

        if let errorParam = errorParam {
            return (false, nil, nil, nil, "Erreur OAuth: \(errorParam)")
        }

        guard let code = code else {
            return (false, nil, nil, nil, "Code d'autorisation non trouvé dans l'URL")
        }

        // Verify state
        let storedState = UserDefaults.standard.string(forKey: WahooStorageKeys.state)
        if state != storedState {
            return (false, nil, nil, nil, "État OAuth invalide. Veuillez réessayer.")
        }

        // Get user ID
        guard let userId = UserDefaults.standard.string(forKey: WahooStorageKeys.userId) else {
            return (false, nil, nil, nil, "ID utilisateur non trouvé")
        }

        // Exchange code for tokens
        return await exchangeCodeForTokens(code: code, userId: userId)
    }

    // MARK: - Exchange Code for Tokens

    func exchangeCodeForTokens(code: String, userId: String) async -> (success: Bool, tokens: WahooTokens?, profile: WahooProfile?, message: String?, error: String?) {
        struct ExchangeRequest: Encodable {
            let code: String
            let userId: String
            let redirectUri: String

            enum CodingKeys: String, CodingKey {
                case code
                case userId = "user_id"
                case redirectUri = "redirect_uri"
            }
        }

        struct ExchangeResponse: Decodable {
            let accessToken: String
            let refreshToken: String
            let expiresIn: Int
            let tokenType: String
            let profile: WahooProfile?

            enum CodingKeys: String, CodingKey {
                case accessToken = "access_token"
                case refreshToken = "refresh_token"
                case expiresIn = "expires_in"
                case tokenType = "token_type"
                case profile
            }
        }

        do {
            let request = ExchangeRequest(
                code: code,
                userId: userId,
                redirectUri: WahooConfig.redirectUri
            )

            let response: ExchangeResponse = try await api.post("/auth/wahoo/exchange", body: request)

            let tokens = WahooTokens(
                accessToken: response.accessToken,
                refreshToken: response.refreshToken,
                expiresIn: response.expiresIn,
                tokenType: response.tokenType
            )

            // Store tokens
            if let data = try? JSONEncoder().encode(tokens) {
                UserDefaults.standard.set(data, forKey: WahooStorageKeys.tokens)
            }

            // Store profile if available
            if let profile = response.profile, let data = try? JSONEncoder().encode(profile) {
                UserDefaults.standard.set(data, forKey: WahooStorageKeys.profile)
            }

            // Clean up temporary data
            UserDefaults.standard.removeObject(forKey: WahooStorageKeys.state)
            UserDefaults.standard.removeObject(forKey: WahooStorageKeys.userId)

            return (true, tokens, response.profile, "Connexion Wahoo établie avec succès!", nil)
        } catch {
            return (false, nil, nil, nil, error.localizedDescription)
        }
    }

    // MARK: - Get Connection Status

    func getConnectionStatus(userId: String) async -> WahooConnectionStatus {
        struct StatusResponse: Decodable {
            let connected: Bool
            let profile: WahooProfile?
            let lastSync: String?

            enum CodingKeys: String, CodingKey {
                case connected
                case profile
                case lastSync = "last_sync"
            }
        }

        do {
            let response: StatusResponse = try await api.get(
                "/auth/wahoo/status",
                queryParams: ["user_id": userId]
            )

            return WahooConnectionStatus(
                isConnected: response.connected,
                profile: response.profile,
                lastSync: response.lastSync,
                error: nil
            )
        } catch {
            // Try to read from local storage
            if let data = UserDefaults.standard.data(forKey: WahooStorageKeys.tokens),
               let _ = try? JSONDecoder().decode(WahooTokens.self, from: data) {

                var profile: WahooProfile? = nil
                if let profileData = UserDefaults.standard.data(forKey: WahooStorageKeys.profile) {
                    profile = try? JSONDecoder().decode(WahooProfile.self, from: profileData)
                }

                return WahooConnectionStatus(
                    isConnected: true,
                    profile: profile,
                    lastSync: nil,
                    error: nil
                )
            }

            return WahooConnectionStatus(
                isConnected: false,
                profile: nil,
                lastSync: nil,
                error: error.localizedDescription
            )
        }
    }

    // MARK: - Disconnect

    func disconnect(userId: String) async -> (success: Bool, error: String?) {
        struct DisconnectResponse: Decodable {
            let message: String
        }

        struct DisconnectRequest: Encodable {
            let userId: String

            enum CodingKeys: String, CodingKey {
                case userId = "user_id"
            }
        }

        do {
            let _: DisconnectResponse = try await api.post(
                "/auth/wahoo/disconnect",
                body: DisconnectRequest(userId: userId)
            )

            // Clean up local storage
            UserDefaults.standard.removeObject(forKey: WahooStorageKeys.tokens)
            UserDefaults.standard.removeObject(forKey: WahooStorageKeys.profile)
            UserDefaults.standard.removeObject(forKey: WahooStorageKeys.state)
            UserDefaults.standard.removeObject(forKey: WahooStorageKeys.userId)

            return (true, nil)
        } catch {
            return (false, error.localizedDescription)
        }
    }

    // MARK: - Sync Workouts

    func syncWorkouts(userId: String) async -> (success: Bool, workouts: [WahooWorkout]?, error: String?) {
        struct SyncRequest: Encodable {
            let userId: String

            enum CodingKeys: String, CodingKey {
                case userId = "user_id"
            }
        }

        struct SyncResponse: Decodable {
            let workouts: [WahooWorkout]
            let count: Int
        }

        do {
            let response: SyncResponse = try await api.post(
                "/auth/wahoo/sync",
                body: SyncRequest(userId: userId)
            )

            return (true, response.workouts, nil)
        } catch {
            return (false, nil, error.localizedDescription)
        }
    }

    // MARK: - Get Recent Workouts

    func getRecentWorkouts(userId: String, limit: Int = 10) async -> (success: Bool, workouts: [WahooWorkout]?, error: String?) {
        struct WorkoutsResponse: Decodable {
            let workouts: [WahooWorkout]
        }

        do {
            let response: WorkoutsResponse = try await api.get(
                "/auth/wahoo/workouts",
                queryParams: [
                    "user_id": userId,
                    "limit": String(limit)
                ]
            )

            return (true, response.workouts, nil)
        } catch {
            return (false, nil, error.localizedDescription)
        }
    }
}
